zookeeper-api/src/main/java/org/apache/helix/zookeeper/routing/RoutingDataManager.java

```diff
@@ -0,0 +1,180 @@
+      throws InvalidRoutingDataException {
+    if (DEFAULT_MSDS_ENDPOINT == null || DEFAULT_MSDS_ENDPOINT.isEmpty()) {
+      throw new IllegalStateException(
+          "HttpRoutingDataReader was unable to find a valid MSDS endpoint String in System "
+              + "Properties!");
+    }
+    return getMetadataStoreRoutingData(RoutingDataReaderType.HTTP, DEFAULT_MSDS_ENDPOINT);
+  }
+
+  /**
+   * Returns the routing data read from MSDS as a MetadataStoreRoutingData object.
+   * @param routingDataReaderType
+   * @param endpoint
+   * @return
+   * @throws IOException
+   * @throws InvalidRoutingDataException
+   */
+  public static MetadataStoreRoutingData getMetadataStoreRoutingData(
+      RoutingDataReaderType routingDataReaderType, String endpoint)
+      throws InvalidRoutingDataException {
+    String routingDataCacheKey = getRoutingDataCacheKey(routingDataReaderType, endpoint);
+    MetadataStoreRoutingData metadataStoreRoutingData =
+        _metadataStoreRoutingDataMap.get(routingDataCacheKey);
+    if (metadataStoreRoutingData == null) {
+      synchronized (RoutingDataManager.class) {
+        metadataStoreRoutingData = _metadataStoreRoutingDataMap.get(routingDataCacheKey);
+        if (metadataStoreRoutingData == null) {
+          metadataStoreRoutingData =
+              new TrieRoutingData(getRawRoutingData(routingDataReaderType, endpoint));
+          _metadataStoreRoutingDataMap.put(routingDataCacheKey, metadataStoreRoutingData);
+        }
+      }
+    }
+    return metadataStoreRoutingData;
+  }
+
+  /**
+   * Clears the statically-cached routing data and private fields.
+   */
+  public synchronized static void reset() {
+    _rawRoutingDataMap.clear();
+    _metadataStoreRoutingDataMap.clear();
+  }
+
+  /**
+   * Returns an appropriate instance of RoutingDataReader given the type.
+   * @param routingDataReaderType
+   * @return
+   */
+  private static RoutingDataReader resolveRoutingDataReader(
+      RoutingDataReaderType routingDataReaderType) {
+    // RoutingDataReaderType.HTTP by default if not found
+    routingDataReaderType =
+        routingDataReaderType == null ? RoutingDataReaderType.HTTP : routingDataReaderType;
+
+    // Instantiate an instance of routing data reader using the type
+    try {
+      return (RoutingDataReader) Class.forName(routingDataReaderType.getClassName()).newInstance();
+    } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
+      throw new MultiZkException(
+          "RoutingDataManager: failed to instantiate RoutingDataReader! ReaderType className: "
+              + routingDataReaderType.getClassName(), e);
+    }
+  }
+
+  /**
+   * Constructs a key for the cache lookup.
+   * @param routingDataReaderType
+   * @param endpoint
+   * @return
+   */
+  private static String getRoutingDataCacheKey(RoutingDataReaderType routingDataReaderType,
+      String endpoint) {
+    return routingDataReaderType.name() + "_" + endpoint;
```

Comment:
If in `resolveRoutingDataReader()` we assume `routingDataReaderType` could be null, it's possible to have a NullPointerExcpetion here.

Reply:
Good comment. `resolveRoutingDataReader`'s check should move to `getRoutingDataCacheKey`, and we shouldn't expect the type to be null. Updated.